storage/innobase/trx/trx0rseg.cc

```diff
@@ -491,7 +491,8 @@ static dberr_t trx_rseg_mem_restore(trx_rseg_t *rseg, trx_id_t &max_trx_id,
 		}
 	}
 
-	if (srv_operation == SRV_OPERATION_RESTORE) {
+	if ((srv_operation == SRV_OPERATION_RESTORE) ||
```

Comment:
Should the correction be added to `is_mariabackup_restore_or_export()`  (storage/innobase/include/srv0srv.h) and for this, and the other case, be changed to use this function?

Reply:
I am not 100% certain whether we should make the change to is_mariabackup_restore_or_export(). Because is_mariabackup_restore_or_export checks whether server operation is SRV_OPERATION_RESTORE or SRV_OPERATION_RESTORE_ROLLBACK_XA or SRV_OPERATION_RESTORE_EXPORT (See [1]). I am not 100% certain whether we should return when server action is SRV_OPERATION_RESTORE_ROLLBACK_XA also.

[1]
inline bool is_mariabackup_restore_or_export()
{
	return is_mariabackup_restore()
	       || srv_operation == SRV_OPERATION_RESTORE_EXPORT;
}

inline bool is_mariabackup_restore()
{
	/* To rollback XA's trx_sys must be initialized, the rest is the same
	as regular backup restore, that is why we join this two operations in
	the most cases. */
	return srv_operation == SRV_OPERATION_RESTORE
	       || srv_operation == SRV_OPERATION_RESTORE_ROLLBACK_XA;
}

Reply:
`mariabackup --prepare --export` is supposed to start up more or less like a normal server. It must restore all undo logs, so that any incomplete transactions in the to-be-exported tables can be rolled back. The suggested changes to `trx_rseg_mem_restore()` and `trx_lists_init_at_db_start()` would seem to break that logic.

Reply:
@dr-m Thank you for the comment. This question maybe due to lack of my understanding; so please help me to understand -- 
1. When are we exactly writing transaction data to the redo log ? is it upon transaction commit or can it happen before transaction commit also ? -- I do understand if the transaction is quite large and involves many index changes then we need to write to the redo log prior to commit to save memory. Also, does this has any connection to the transaction isolation level and autocommit statements ?
2. If we are writing transaction data to the redo log prior to transaction commit, then is there a separate redo log event to identify the transaction commit (Something like MLOG_COMMIT; couldnt find such MLOG event in 10.2 codebase) ?

Assuming we write to the redo log prior to transaction commit, I think I agree with your comment and let me think about an alternative fix keeping what you said in my head.

Thanks
Thejaka

Reply:
Fixed.

Reply:
I am sorry for missing your replies. You may be on the right track, but the code has changed a lot, mainly in the 10.6 branch. I would like to see also a 10.6 version of this. You’d better avoid any unnecessary white space changes to ease merging or cherry-picking.

1. Transaction data is written to undo log pages (rollback segment header pages and undo log pages). Starting with MariaDB Server 10.3, the `TRX_SYS` page is a read-only directory pointing to the rollback segment header pages. Any persistent InnoDB data pages are covered by a write-ahead log (redo log, `ib_logfile0`).
1. Undo log records are indeed written before any data modifications. On transaction commit, only the transaction state in the undo log header will be updated.

You may find this old presentation ([slides](https://docs.google.com/presentation/d/1bP3yh57B58yfGDd-34TZ5MeiacIdaseaTLpd_Ov5NI8/) and [video](https://vimeo.com/258533150)) useful.